lib/spack/spack/util/environment.py

```diff
@@ -919,8 +918,14 @@ def _source_single_file(file_and_args, environment):
         source_file.extend(x for x in file_and_args)
         source_file = ' '.join(source_file)
 
-        dump_environment = 'PYTHONHOME="{0}" "{1}" -c "{2}"'.format(
-            sys.prefix, sys.executable, py_cmd)
+        # If the environment contains 'python' use it, if not
+        # go with sys.executable. Below we just need a working
+        # Python interpreter, not necessarily sys.executable.
+        python_command = executable.which('python')
```

Comment:
This should be `which('python3', 'python', 'python2')`

Reply:
@adamjstewart @tgamblin I've been thinking a bit about the order in this context. Probably the best one here is `which('python', 'python3', 'python2')`. The rationale is that the order above is made to prefer Python3, while here we should prefer "the first working Python in PATH". The order we have right now will for instance fail if somebody is:

- Using Spack with a Python 3.X interpreter
- Building a DAG that uses a Python 2.X interpreter and requires `_source_single_file`

In that case we'll have `PYTHONHOME` set for Python 2 and Python 3 will be found first. If we had `python` as a first match, the failure would occur in possibly rarer cases like people that builds `python~pythoncmd`.

As Todd commented below, there's nothing 100% correct if we go this way and we need a more robust solution overall (maybe recording at the beginning all the details of the Python interpreter Spack is started with?), I just wanted to share this thought. Let me know if you want a one line fix submitted for this, or if we should wait for a bug report and act on that.